scene/3d/mesh_instance.cpp

```diff
@@ -163,15 +172,242 @@ void MeshInstance::_resolve_skeleton_path() {
 		}
 	}
 
+	if (new_skin_reference == skin_ref) {
+		return;
+	}
+
+	if (skin_ref.is_valid() && mesh.is_valid() && _is_software_skinning_enabled()) {
+		ERR_FAIL_COND(!skin_ref->get_skeleton_node());
+		skin_ref->get_skeleton_node()->disconnect("skeleton_updated", this, "_update_skinning");
+	}
+
 	skin_ref = new_skin_reference;
 
-	if (skin_ref.is_valid()) {
-		VisualServer::get_singleton()->instance_attach_skeleton(get_instance(), skin_ref->get_skeleton());
+	software_skinning_bones_ready = false;
+
+	_initialize_skinning();
+}
+
+bool MeshInstance::_is_global_software_skinning_enabled() {
+	// Check if forced in project settings.
+	if (GLOBAL_GET("rendering/quality/skinning/force_software_skinning")) {
+		return true;
+	}
+
+	// Check if enabled in project settings.
+	if (!GLOBAL_GET("rendering/quality/skinning/software_skinning_fallback")) {
+		return false;
+	}
+
+	// Check if requested by renderer settings.
+	return VSG::storage->has_os_feature("skinning_fallback");
+}
+
+bool MeshInstance::_is_software_skinning_enabled() const {
+	static bool global_software_skinning = _is_global_software_skinning_enabled();
+	return global_software_skinning;
+}
```

Comment:
I need the whole logic to be executed when initializing the local static bool to make sure it's done only once, so in order to keep things readable I separated the code into a static method. But I don't mind setting it up with a combination of logical operators directly in `_is_software_skinning_enabled` instead if you prefer.

Reply:
I think what @clayjohn is getting at is that having the bool static here doesn't seem to achieve the aim of only setting global_software_skinning only once (in fact it pretty much seems a no-op, as I don't think you can access static local variables from outside a function). If you wanted to set a static afaik it would be better a static member variable, which you can access from elsewhere.

Reply:
The goal is for `_is_global_software_skinning_enabled` to be called only once and it does that for me (with VS 2019). Do you get a different result with your compiler?

Here's the c++ reference I used to confirm how initialization works for static local variables:
https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables

Reply:
Ah I was not familiar with that method. It may do the job, as long as it is first called in the right order, i.e. after the godot GLOBALS have been defined. It may be better to explicitly set it at a fixed time to prevent obscure order of construction bugs - see what clayjohn thinks on this.

Reply:
Sounds fine to me. I was unfamiliar with the behaviour of static local variables as well.

Reply:
One thing it might be worth doing is fully commenting this behaviour. Given that both myself and clayjohn didn't spot what it was doing :blush: , it is likely that we might make the same mistake coming back to this code later, or another programmer reading it.

Reply:
Good point, I'll add comments.